prov/rxm/src/rxm_ep.c

```diff
@@ -1423,65 +1423,20 @@ static int rxm_listener_close(struct rxm_ep *rxm_ep)
 
 static int rxm_ep_close(struct fid *fid)
 {
-	struct rxm_ep *rxm_ep;
 	int ret, retv = 0;
+	struct rxm_ep *rxm_ep =
+		container_of(fid, struct rxm_ep, util_ep.ep_fid.fid);
+	struct rxm_ep_wait_ref *wait_ref;
 
-	rxm_ep = container_of(fid, struct rxm_ep, util_ep.ep_fid.fid);
-
-	if (rxm_ep->util_ep.tx_cq->wait) {
-		ret = ofi_wait_fd_del(rxm_ep->util_ep.tx_cq->wait,
-				      rxm_ep->msg_cq_fd);
-		if (ret)
-			retv = ret;
-	}
-
-	if (rxm_ep->util_ep.rx_cq->wait) {
-		ret = ofi_wait_fd_del(rxm_ep->util_ep.rx_cq->wait,
-				      rxm_ep->msg_cq_fd);
-		if (ret)
-			retv = ret;
-	}
-
-	if (rxm_ep->util_ep.tx_cntr && rxm_ep->util_ep.tx_cntr->wait) {
-		ret = ofi_wait_fd_del(rxm_ep->util_ep.tx_cntr->wait,
-				      rxm_ep->msg_cq_fd);
-		if (ret)
-			retv = ret;
-	}
-
-	if (rxm_ep->util_ep.rx_cntr && rxm_ep->util_ep.rx_cntr->wait) {
-		ret = ofi_wait_fd_del(rxm_ep->util_ep.rx_cntr->wait,
-				      rxm_ep->msg_cq_fd);
-		if (ret)
-			retv = ret;
-	}
-
-	if (rxm_ep->util_ep.rd_cntr && rxm_ep->util_ep.rd_cntr->wait) {
-		ret = ofi_wait_fd_del(rxm_ep->util_ep.rd_cntr->wait,
-				      rxm_ep->msg_cq_fd);
-		if (ret)
-			retv = ret;
-	}
-
-	if (rxm_ep->util_ep.wr_cntr && rxm_ep->util_ep.wr_cntr->wait) {
-		ret = ofi_wait_fd_del(rxm_ep->util_ep.wr_cntr->wait,
-				      rxm_ep->msg_cq_fd);
-		if (ret)
-			retv = ret;
-	}
-
-	if (rxm_ep->util_ep.rem_rd_cntr && rxm_ep->util_ep.rem_rd_cntr->wait) {
-		ret = ofi_wait_fd_del(rxm_ep->util_ep.rem_rd_cntr->wait,
-				      rxm_ep->msg_cq_fd);
-		if (ret)
-			retv = ret;
-	}
-
-	if (rxm_ep->util_ep.rem_wr_cntr && rxm_ep->util_ep.rem_wr_cntr->wait) {
-		ret = ofi_wait_fd_del(rxm_ep->util_ep.rem_wr_cntr->wait,
+	dlist_foreach_container(&rxm_ep->msg_cq_fd_ref_list,
+				struct rxm_ep_wait_ref,
+				wait_ref, entry) {
+		ret = ofi_wait_fd_del(wait_ref->wait,
 				      rxm_ep->msg_cq_fd);
```

Comment:
IIUC, this still deletes the fd multiple times since we add them all to the msg_cq_fd_ref_list without checking if it's already present. 
Also we have ref counting for fd in util_wait. Isn't that sufficient to handle multiple additions/deletions?

Reply:
This particular change is walking a list and deleting each item from it.

Reply:
Ah ok. So this is fixing multiple deletion of the wait object and not the fd. I was looking at the commit message and it said otherwise. Reading the PR description comment makes sense.

@gladkovdmitry17 can you please fix the commit message title and add the details?

Reply:
Sure, I have updated commit message and PR titles.